apps/finance/app/src/components/Balances.js

```diff
@@ -1,99 +1,119 @@
+  const retryDelay = useRef(CONVERT_API_RETRY_DELAY)
+
+  const symbolsQuery = symbols.join(',')
+
+  useEffect(() => {
+    let cancelled = false
+    let retryTimer = null
+
+    const update = async () => {
+      if (!symbolsQuery) {
+        setRates({})
+        return
+      }
 
-    const res = await fetch(convertApiUrl(verifiedSymbols))
-    const convertRates = await res.json()
-    this.setState({ convertRates })
-  }, CONVERT_THROTTLE_TIME)
-
-  render() {
-    const { compactMode, balances, theme } = this.props
-    const { convertRates } = this.state
-    const balanceItems = balances.map(
-      ({ address, numData: { amount, decimals }, symbol, verified }) => {
-        const adjustedAmount = amount / Math.pow(10, decimals)
-        const convertedAmount =
-          verified && convertRates[symbol]
-            ? adjustedAmount / convertRates[symbol]
-            : -1
-        return {
-          address,
-          symbol,
-          verified,
-          amount: round(adjustedAmount, 5),
-          convertedAmount: round(convertedAmount, 5),
+      try {
+        const response = await fetch(convertRatesUrl(symbolsQuery))
+        const rates = await response.json()
+        if (!cancelled) {
+          setRates(rates)
+          retryDelay.current = CONVERT_API_RETRY_DELAY
+        }
+      } catch (err) {
+        // The !cancelled check is needed in case:
+        //  1. The fetch() request is ongoing.
+        //  2. The component gets unmounted.
+        //  3. An error gets thrown.
+        //
+        //  Assuming the fetch() request keeps throwing, it would create new
+        //  requests even though the useEffect() got cancelled.
+        if (!cancelled) {
+          // Add more delay after every failed attempt
+          retryDelay.current = Math.min(
+            CONVERT_API_RETRY_DELAY_MAX,
+            retryDelay.current * 1.2
+          )
+          retryTimer = setTimeout(update, retryDelay.current)
         }
       }
-    )
-    const noBalances = balanceItems.length === 0
+    }
+    update()
+
+    return () => {
+      cancelled = true
+      clearTimeout(retryTimer)
+      retryDelay.current = CONVERT_API_RETRY_DELAY
+    }
+  }, [symbolsQuery])
+
+  return rates
+}
 
-    return (
-      <Box heading="Token Balances">
+// Prepare the balances for the BalanceToken component
+function useBalanceItems(balances) {
+  const verifiedSymbols = balances
+    .filter(({ verified }) => verified)
+    .map(({ symbol }) => symbol)
+
+  const convertRates = useConvertRates(verifiedSymbols)
+
+  const balanceItems = useMemo(() => {
+    return balances.map(({ address, amount, decimals, symbol, verified }) => ({
+      address,
+      amount,
+      convertedAmount: convertRates[symbol]
+        ? amount.divn(convertRates[symbol])
+        : new BN(-1),
+      decimals,
+      symbol,
+      verified,
+    }))
+  }, [balances, convertRates])
+
+  return balanceItems
+}
+
+function Balances({ balances, compactMode }) {
```

Comment:
Not a big deal, but if we're using `useLayout()` here anyway, should we avoid drilling `compactMode` down?

Reply:
Yes! We can even remove it from `App` entirely: https://github.com/aragon/aragon-apps/pull/1121/commits/5de1e668a1e15b2cd69478a76b5566206b251f43